Sources/SignalProducer.swift

```diff
@@ -1877,6 +1877,59 @@ extension SignalProducerProtocol {
 			}
 		}
 	}
+	
+	/// Creates a new `SignalProducer` that will bucket each received value into
+	/// a group based on the key returned from `grouping`. 
+	///
+	/// Termination events on the original signal are also forwarded to each
+	/// producer group.
+	///
+	/// - parameters:
+	///   - grouping: a closure that determines the grouping key for a given value
+	///
+	/// - returns: A producer of producers amits one producer for each group and forwards
+	///            each value from the original producer to the inner producer corresponding 
+	///            to the group to which the value belongs to (as determined by the key)
+	public func group<Key: Hashable>(by grouping: @escaping (Value) -> Key) -> SignalProducer<(Key, SignalProducer<Value, Error>), Error> {
+		return SignalProducer { observer, disposable in
+			let groups = Atomic<[Key: Signal<Value, Error>.Observer]>([:])
+			
+			disposable += self.start { event in
+				switch event {
+				case let .value(value):
+					let key = grouping(value)
+					let group: Signal<Value, Error>.Observer = groups.modify { groups in
+						if let group = groups[key] {
+							return group
+						} else {
+							let (signal, innerObserver) = Signal<Value, Error>.pipe()
+							let producer = SignalProducer(signal).replayLazily(upTo: Int.max)
```

Comment:
I _think_ so, otherwise consumers could miss events without realizing it. Hence why I erred on the side of buffering. I could be convinced otherwise, but I'm not sure how to best communicate those semantics at the operator level.